tortoise/state.go

```diff
@@ -0,0 +1,176 @@
+package tortoise
+
+import (
+	"bytes"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/common/util"
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+const (
+	namespaceLast     = "l"
+	namespaceEvict    = "e"
+	namespaceVerified = "v"
+	namespaceGood     = "g"
+	namespaceOpinons  = "o"
+)
+
+type state struct {
+	log log.Log
+	db  database.Database
+
+	// if true only non-flushed items will be synced
+	diffMode bool
+
+	Last            types.LayerID
+	EvictFrom       types.LayerID
+	Verified        types.LayerID
+	GoodBlocksIndex map[types.BlockID]bool
+	// use 2D array to be able to iterate from latest elements easily
+	BlockOpinionsByLayer map[types.LayerID]map[types.BlockID]Opinion // records hdist, for each block, its votes about every
+	// TODO: Tal says: We keep a vector containing our vote totals (positive and negative) for every previous block
+	// that's not needed here, probably for self healing?
+}
+
+func (s *state) Persist() error {
+	batch := s.db.NewBatch()
+
+	if err := batch.Put([]byte(namespaceLast), s.Last.Bytes()); err != nil {
+		return err
+	}
+	if err := batch.Put([]byte(namespaceEvict), s.EvictFrom.Bytes()); err != nil {
+		return err
+	}
+	if err := batch.Put([]byte(namespaceVerified), s.Verified.Bytes()); err != nil {
+		return err
+	}
+	var b bytes.Buffer
+	for id, flushed := range s.GoodBlocksIndex {
+		if flushed && s.diffMode {
+			continue
+		}
+		s.GoodBlocksIndex[id] = true
+		b.WriteString(namespaceGood)
+		b.Write(id.Bytes())
+		if err := batch.Put(b.Bytes(), nil); err != nil {
+			return err
+		}
+		b.Reset()
+	}
+
+	for layer, blocks := range s.BlockOpinionsByLayer {
+		for block1, opinions := range blocks {
+			for block2, opinion := range opinions {
+				if opinion.Flushed && s.diffMode {
+					continue
+				}
+				opinion.Flushed = true
```

Comment:
mark as flushed after `batch.Put` succeeds

Reply:
it doesn't matter. actual write will happen in batch.Write